Test failing sporadically (test_annotations_per_doc_not_enforced_for_training_or_test)

Recently one of the Python tests in particular has been failing sporadically when run in the GitHub workflows.  The scenario for this particular test is a project with:

- 5 training, 10 test and 20 annotation documents
- 6 annotators
- 3 annotations per document
- each annotator max proportion 0.6 (i.e. no one person should be allowed to annotate more than 60% of the corpus)

The test has each annotator in turn _attempt_ to complete all 5 training, 10 test, and 20 annotation documents, and checks that

1. all 6 annotators are allowed to complete all 5 training docs and all 10 test docs
  - i.e. the 60% proportion is not applied to training or test
2. no annotator was allowed to complete _more_ than 60% (12) of the 20 annotation documents
  - i.e. the 60% _does_ apply to annotation
3. the total number of actual annotations completed on the main 20 annotation docs adds up to 60
  - three annotators per document times 20 documents
4. at least one annotator was given strictly less than 12 documents
  - this proves that the three-annotations-per-document limit trumps the 60%-per-annotator.

However the test is sporadically failing at step 3, finding that there are only 59 annotations in total instead of the expected 50.  This only fails some of the time but when it does fail it's always 59 rather than 60.

There must be some weird interaction between the way we randomise the order in which documents are handed to the annotators vs exactly how we calculate the 60% threshold - will probably need to run the test a few hundred times recording the exact order in which documents are handed out and see if we can spot any patterns.

This test is also failing sporadically, I'll investigate it in this PR together:

```
=================================== FAILURES ===================================
______________ TestProjectModel.test_get_annotator_document_score ______________

self = <backend.tests.test_models.TestProjectModel testMethod=test_get_annotator_document_score>

    def test_get_annotator_document_score(self):
        annotator = self.annotators[0]
        # No annotations == 0
        self.assertEqual(0, self.project.get_annotator_document_score(annotator, DocumentType.TEST))
    
        incorrect_data = {
            "label1": "Incorrect",
            "label2": "Incorrect"
        }
    
        # All incorrect
        for doc in self.test_docs:
            anno = Annotation.objects.create(user=annotator, document=doc)
            anno.data = incorrect_data
    
        self.assertEqual(0, self.project.get_annotator_document_score(annotator, DocumentType.TEST))
    
        # All correct
        annotator2 = self.annotators[1]
        for doc in self.test_docs:
            correct_annotation_data = {
                "label1": doc.data["gold"]["label1"]["value"],
                "label2": doc.data["gold"]["label2"]["value"],
            }
            anno = Annotation.objects.create(user=annotator2, document=doc)
            anno.data = correct_annotation_data
    
>       self.assertEqual(self.num_test_docs, self.project.get_annotator_document_score(annotator2, DocumentType.TEST))
E       AssertionError: 7 != 0

backend/tests/test_models.py:696: AssertionError
```

--- a/backend/models.py
+++ b/backend/models.py
@@ -737,9 +737,18 @@ class Project(models.Model):
         Annotation task performs an extra check for remaining annotation task (num_annotation_tasks_remaining),
         testing and training does not do this check as the annotator must annotate all documents.
         """
-        if (DocumentType.ANNOTATION and self.num_annotation_tasks_remaining > 0) or \
-                DocumentType.TEST or DocumentType.TRAINING:
-            for doc in self.documents.filter(doc_type=doc_type).order_by('?'):
+        if (doc_type == DocumentType.ANNOTATION and self.num_annotation_tasks_remaining > 0) or \
+                doc_type == DocumentType.TEST or doc_type == DocumentType.TRAINING:
+            if doc_type == DocumentType.TEST or doc_type == DocumentType.TRAINING:
+                queryset = self.documents.filter(doc_type=doc_type).order_by('?')
+            else:
+                # Prefer documents which have fewer complete or pending annotations, in order to
+                # spread the annotators as evenly as possible across the available documents
+                queryset = self.documents.filter(doc_type=doc_type).alias(
+                    occupied_annotations=Count("annotations", filter=Q(annotations__status=Annotation.COMPLETED)
+                                                                     | Q(annotations__status=Annotation.PENDING))
+                ).order_by('occupied_annotations', '?')
+            for doc in queryset:
                 # Check that annotator hasn't annotated and that
                 # doc hasn't been fully annotated
                 if doc.user_can_annotate_document(user):

--- a/backend/tests/test_rpc_endpoints.py
+++ b/backend/tests/test_rpc_endpoints.py
@@ -8,6 +8,7 @@ from django.test import TestCase, Client
 
 from django.utils import timezone
 import json
+import logging
 
 from backend.models import Annotation, Document, DocumentType, Project, AnnotatorProject, UserDocumentFormatPreference
 from backend.rpc import create_project, update_project, add_project_document, add_document_annotation, \
@@ -28,7 +29,7 @@ from backend.tests.test_models import create_each_annotation_status_for_user, Te
 from backend.tests.test_rpc_server import TestEndpoint
 
 
-
+LOGGER = logging.getLogger(__name__)
 
 class TestUserAuth(TestCase):
 
@@ -1665,14 +1666,29 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
 
         # Expect to get self.num_training_docs tasks
         num_completed_tasks = 0
+        if expected_doc_type_str == 'Annotation':
+            all_docs = self.docs
+        elif expected_doc_type_str == 'Training':
+            all_docs = self.training_docs
+        else:
+            all_docs = self.test_docs
+
+        annotated_docs = {doc.pk: ' ' for doc in all_docs}
         for i in range(num_annotations_to_complete):
             task_context = get_annotation_task(ann_req)
             if task_context:
                 self.assertEqual(expected_doc_type_str, task_context.get("document_type"),
                                  f"Document type does not match in task {task_context!r}, " +
                                  "annotator {ann.username}, document {i}")
+                annotated_docs[task_context['document_id']] = "\u2714"
                 complete_annotation_task(ann_req, task_context["annotation_id"], {"sentiment": answer})
                 num_completed_tasks += 1
+
+        # Draw a nice markdown table of exactly which documents each annotator was given
+        if annotator == 0:
+            LOGGER.debug("Annotator | " + (" | ".join(str(i) for i in annotated_docs.keys())))
+            LOGGER.debug(" | ".join(["--"] * (len(annotated_docs)+1)))
+        LOGGER.debug(ann.username + " | " + (" | ".join(str(v) for v in annotated_docs.values())))
 
         return num_completed_tasks
 

--- a/backend/tests/test_rpc_endpoints.py
+++ b/backend/tests/test_rpc_endpoints.py
@@ -1609,10 +1609,11 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
         self.proj.save()
 
         docs_annotated_per_user = []
-        for (i, (ann_user, _)) in enumerate(self.annotators):
+        for (ann_user, _) in self.annotators:
             # Add to project
             self.assertTrue(add_project_annotator(self.manager_request, self.proj.id, ann_user.username))
 
+        for (i, (ann_user, _)) in enumerate(self.annotators):
             # Every annotator should be able to complete every training document, even though
             # max annotations per document is less than the total number of annotators
             self.assertEqual(self.num_training_docs,
@@ -1623,6 +1624,7 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
             self.assertEqual(self.num_training_docs,
                              self.proj.get_annotator_document_score(ann_user, DocumentType.TRAINING))
 
+        for (i, (ann_user, _)) in enumerate(self.annotators):
             # Every annotator should be able to complete every test document, even though
             # max annotations per document is less than the total number of annotators
             self.assertEqual(self.num_test_docs,
@@ -1633,6 +1635,7 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
             self.assertEqual(self.num_training_docs,
                              self.proj.get_annotator_document_score(ann_user, DocumentType.TRAINING))
 
+        for (i, (ann_user, _)) in enumerate(self.annotators):
             # Now attempt to complete task normally
             num_annotated = self.complete_annotations(self.num_docs, "Annotation", annotator=i)
             docs_annotated_per_user.append(num_annotated)

--- a/backend/tests/test_rpc_endpoints.py
+++ b/backend/tests/test_rpc_endpoints.py
@@ -1379,7 +1379,7 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
         self.num_training_docs = 5
         self.training_docs = []
         for i in range(self.num_training_docs):
-            self.docs.append(Document.objects.create(project=self.proj,
+            self.training_docs.append(Document.objects.create(project=self.proj,
                                                      doc_type=DocumentType.TRAINING,
                                                      data={
                                                         "text": f"Document {i}",
@@ -1396,7 +1396,7 @@ class TestAnnotationTaskManagerTrainTestMode(TestEndpoint):
         self.num_test_docs = 10
         self.test_docs = []
         for i in range(self.num_test_docs):
-            self.docs.append(Document.objects.create(project=self.proj,
+            self.test_docs.append(Document.objects.create(project=self.proj,
                                                      doc_type=DocumentType.TEST,
                                                      data={
                                                          "text": f"Document {i}",

